### PR TITLE
ModelFactory for dynamic model dependency requirements

### PR DESCRIPTION
**Pull request by alanhartless**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | n
| New feature? | y - dev
| Related user documentation PR URL | na
| Related developer documentation PR URL | todo
| Issues addressed (#s or URLs) | na
| BC breaks? | n
| Deprecations? | n

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

Since the deprecation of MauticFactory, there are scenarios where it is unknown what models are required and thus dependency injection is impossible. For example, where we have stored channel/source (that correlates with models) in tables. Page hits for example make it possible to get the entity's name by doing something like `$factory->getModel($channel)->getEntity($channelId)`. This PR simply inherits that functionality intended to be used only for such a scenario. This is not meant to "save time" compared to injecting independent models as needed (or we've recreated MauticFactory all over again). 

#### Steps to test this PR:
1. Quick means is to manually edit a page_hit entry injecting the columns source/source_id with `email` and an ID of an existing Mautic email ID. 
2. View the timeline of the contact for that page hit - the name of the email should be noted as teh source.

### CODE REVIEW

**Comment by mqueme** [comment]
Can this be used in all places where channel and channel_id are now being implemented?

**Comment by alanhartless** [comment]
Sure - if it's needed to get the entity object of the channel/channel_id so you can display the name or some other information. Perfect use case.

File: app/bundles/LeadBundle/EventListener/LeadSubscriber.php [hunk 1]
```diff
@@ -529,18 +530,32 @@ protected function addTimelineDoNotContactEntries(Events\LeadTimelineEvent $even
 
                 if (!empty($row['channel'])) {
                     if ($channelModel = $this->getChannelModel($row['channel'])) {
+                        if ($channelModel instanceof ChannelTimelineInterface) {
```
Comment (dongilbert) [inline app/bundles/LeadBundle/EventListener/LeadSubscriber.php]: I see you're doing a check against `ChannelTimelineInterface` to see if the model implements it, but you didn't actually implement that interface on the appropriate classes (`Mautic\LeadBundle\EventListener\LeadSubscriber` and `Mautic\PageBundle\EventListener\LeadSubscriber`)

Comment (alanhartless) [inline app/bundles/LeadBundle/EventListener/LeadSubscriber.php]: Those interfaces are meant to be used on models for 3rd party - but aren't used actively by anything in Mautic.

**Comment by escopecz** [comment]
:+1:

**Pull request merged**